cli/src/cmd/watch.rs

```diff
@@ -0,0 +1,362 @@
+///
+/// **Note** this is a bit weird since we're installing the hook on the config that's already used
+/// in `Watchexec` but necessary if we want to have access to it in order to
+/// [`Watchexec::reconfigure`]
+fn on_action<F, T>(
+    args: WatchArgs,
+    mut config: RuntimeConfig,
+    wx: Arc<Watchexec>,
+    cmd: Vec<String>,
+    other: T,
+    f: F,
+) where
+    F: for<'a> Fn(OnActionState<'a, T>) + Send + 'static,
+    T: Clone + Send + 'static,
+{
+    let on_busy = if args.no_restart { "do-nothing" } else { "restart" };
+    let print_events = args.why;
+    let runtime = config.clone();
+    let w = Arc::clone(&wx);
+    config.on_action(move |action: Action| {
+        let fut = async { Ok::<(), Infallible>(()) };
+        if print_events {
+            for (n, event) in action.events.iter().enumerate() {
+                eprintln!("[EVENT {}] {}", n, event);
+            }
+        }
+
+        let signals: Vec<MainSignal> = action.events.iter().flat_map(|e| e.signals()).collect();
+        let has_paths = action.events.iter().flat_map(|e| e.paths()).next().is_some();
+
+        if signals.contains(&MainSignal::Terminate) {
+            action.outcome(Outcome::both(Outcome::Stop, Outcome::Exit));
+            return fut
+        }
+
+        if signals.contains(&MainSignal::Interrupt) {
+            action.outcome(Outcome::both(Outcome::Stop, Outcome::Exit));
+            return fut
+        }
+
+        if !has_paths {
+            if !signals.is_empty() {
+                let mut out = Outcome::DoNothing;
+                for sig in signals {
+                    out = Outcome::both(out, Outcome::Signal(sig.into()));
+                }
+
+                action.outcome(out);
+                return fut
+            }
+
+            let completion = action.events.iter().flat_map(|e| e.completions()).next();
+            if let Some(status) = completion {
+                let (msg, printit) = match status {
+                    Some(ProcessEnd::ExitError(code)) => {
+                        (format!("Command exited with {}", code), true)
+                    }
+                    Some(ProcessEnd::ExitSignal(sig)) => {
+                        (format!("Command killed by {:?}", sig), true)
+                    }
+                    Some(ProcessEnd::ExitStop(sig)) => {
+                        (format!("Command stopped by {:?}", sig), true)
+                    }
+                    Some(ProcessEnd::Continued) => ("Command continued".to_string(), true),
+                    Some(ProcessEnd::Exception(ex)) => {
+                        (format!("Command ended by exception {:#x}", ex), true)
+                    }
+                    Some(ProcessEnd::Success) => ("Command was successful".to_string(), false),
+                    None => ("Command completed".to_string(), false),
+                };
+
+                if printit {
+                    eprintln!("[[{}]]", msg);
+                }
```

Comment:
Maybe no need to print these? Could perhaps but them behind `tracing::trace`? Seems to clutter command output a bit